src/hooks.js

```diff
@@ -0,0 +1,44 @@
+import { useState, useEffect } from 'preact/hooks'
+
+const fetchBody = async (url, fetchOpts) => {
+  const state = {}
+  try {
+    state.response = await fetch(url, fetchOpts)
+    if (state.response.status !== 204) {
+      state.body = await state.response.text()
+    }
+    state.error = state.response.ok
+      ? undefined
+      : Error(state.response.statusText)
+  } catch (error) {
+    state.error = error
+  }
+  return state
+}
+
+export const useFetchJSON = (...args) => {
+  const state = useFetch(...args)
+  if (!state.body) return state
+  try {
+    state.data = JSON.parse(state.body)
+  } catch (error) {
+    state.error = error
+  }
+  return state
+}
+
+export const useFetch = (url, fetchOpts, inputs = [url]) => {
+  const [state, setState] = useState({ pending: true })
+  // biome-ignore lint/correctness/useExhaustiveDependencies: dependencies defined by the caller
+  useEffect(() => {
+    const controller = new AbortController()
```

Comment:
`fetch` is a promise based API but we can still cancel them, if our compent is removed or changed before the request is over, we need to prevent updating the dom, also if the request is taking time we may even cancel it and save some ressources.

To manage that we can use `AbortController` that have an `.abort()` method and a state of if the request was aborted